README.md
## Laravel Echo for Flutter

Basically, this package is port of official [Laravel Echo javascript library](https://github.com/laravel/echo). It helps subscribe to channels and listen for events broadcasted from your Laravel app.

API is same as official Echo package, so everything in [Official documentation](https://laravel.com/docs/5.7/broadcasting) should work.

Three connectors available:

- [x] [socket.io](#socket.io)
- [x] [Pusher](#pusher)
- [x] Null

<img width="300" alt="Screen Shot 2019-03-17 at 9 37 50 PM" src="https://user-images.githubusercontent.com/7093483/54494522-f15eef80-48fc-11e9-8fc1-e986bc004360.png">

## Getting started

### socket.io

To use with `socket.io`, you need to install [socket_io_client](https://pub.dartlang.org/packages/socket_io_client) for your Flutter app.

In your `pubspec.yaml` file:

```yaml
dependencies:
  ...
  socket_io_client: ^0.9.1
  flutter_laravel_echo:
```

import `socket_io_client`

```dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
```

usage

```dart
// Create echo instance
Echo echo = new Echo({
  'broadcaster': 'socket.io',
  'client': IO.io,
});

// Listening public channel
echo.channel('public-channel').listen('PublicEvent', (e) {
  print(e);
});

// Listening private channel
// Needs auth. See details how to authorize channel below in guides
echo.private('private-channel').listen('PrivateEvent', (e) {
  print(e);
});

// Listening presence channel
// Needs auth. See details how to authorize channel below in guides
echo.join('presence-channel')
  .here((users) {
    print(users);
  }).joining((user) {
    print(user);
  }).leaving((user) {
    print(user);
  }).listen('PresenceEvent', (e) {
    print(e);
  });

// Accessing socket instance
echo.socket.on('connect', (_) => print('connected'));
echo.socket.on('disconnect', (_) => print('disconnected'));
```

### Pusher

To use with **Pusher**, you need to install [flutter_pusher_client](https://pub.dartlang.org/packages/flutter_pusher_client) for you Flutter app.

In your `pubspec.yaml` file:

```yaml
dependencies:
  ...
  flutter_pusher_client: ^0.3.1
  flutter_laravel_echo: ^0.2.5
```

import `flutter_pusher_client`

```dart
import 'package:flutter_pusher_client/flutter_pusher.dart';
```

```dart
PusherOptions options = PusherOptions(
  host: '10.0.2.2',
  port: 6001,
  encrypted: false,
);
FlutterPusher pusher = FlutterPusher('app', options, enableLogging: true);

Echo echo = new Echo({
  'broadcaster': 'pusher',
  'client': pusher,
});

echo.channel('public-channel').listen('PublicEvent', (e) {
  print(e);
});

socket.on('connect', (_) => print('connect'));
socket.on('disconnect', (_) => print('disconnect'));
```

## Guide

### Options

| Option       | Description                                | Default               |
| ------------ | ------------------------------------------ | --------------------- |
| auth         |                                            |                       |
| authEndpoint |                                            | /broadcasting/auth    |
| broadcaster  |                                            | socket.io             |
| crsfToken    |                                            |                       |
| host         | Socket host                                | http://localhost:6001 |
| namespace    | Event namespace                            | App.Events            |
| ...          | Any other options, passed as socket params |                       |

### Authorize private channels

To authorize channel requests we use [Laravel Passport](https://laravel.com/docs/5.7/passport)

In our `BroadcastServiceProvider.php` we need to enable

```php
Broadcast::routes(['middleware' => ['auth:api']]);
```

then, when creating `Echo` instance include `Authorization` header with bearer token

```dart
echo = new Echo({
  'broadcaster': 'socket.io',
  'client': socket,
  'auth': {
    'headers': {
        'Authorization': 'Bearer $token'
    }
  }
});
```

### Sample backend

Backend used for example app could be found at [echo-server](https://github.com/kakajansh/echo-server)

Package by [Kakajan SH](http://kakajan.sh)


CHANGELOG.md
## [0.3.0] - 13.01.2021

- Fix presence event

## [0.2.9] - 13.01.2021

- Update package name

## [0.2.8] - 31.10.2020

- Fix issue with PrivateChannel #33

## [0.2.6] - 27.10.2020

- Update example app
- Add [echo-server](https://github.com/kakajansh/echo-server) sample backend

## [0.2.5] - 27.10.2020

Update to make inline with official Laravel echo package. Including:

- Add stopWhisper method to channel
- Add support for custom connectors
- Add pusher private-encrypted
- Implement error handling with support for Pusher
- Register subscription succeeded callbacks

## [0.2.2] - 15.01.2020

- Added Pusher Client Support by [@heywhy](http://github.com/heywhy/)

## [0.2.1] - 17.03.2018

- Update package description
- Update README.md

## [0.2.0] - 16.03.2018

- Minor changes
- Include example

## [0.1.1] - 02.03.2018

- Added Pusher connector. Implemented by [@abrahamSN](https://github.com/abrahamSN)
- Fixes issue with socket-io private channels
- Cleaned up

## [0.0.1] - 09.02.2018

- Initial release.
- Added null connector
- Added socketio connector


example/lib/pusher.dart
import 'package:example/modal.dart';
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter_laravel_echo/flutter_laravel_echo.dart';
import 'package:flutter_pusher_client/flutter_pusher.dart';

class PusherPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _PusherPage();
}

class _PusherPage extends State<PusherPage> {
  List<String> _logs = new List();
  Echo echo;
  bool isConnected = false;
  String channelType = 'public';
  String channelName = '';
  String event;
  FlutterPusher pusherClient;

  void onConnectionStateChange(ConnectionStateChange event) {
    if (event.currentState == 'CONNECTED') {
      log('connected');
      setState(() {
        isConnected = true;
      });
    } else if (event.currentState == 'DISCONNECTED') {
      log('disconnected');
      setState(() {
        isConnected = false;
      });
    }
  }

  FlutterPusher getPusherClient() {
    PusherOptions options = PusherOptions(
      host: '10.0.2.2',
      port: 6001,
      encrypted: false,
    );
    return FlutterPusher('app', options, lazyConnect: true);
  }

  void connect() {
    pusherClient.connect(onConnectionStateChange: this.onConnectionStateChange);
  }

  @override
  void initState() {
    super.initState();

    pusherClient = getPusherClient();

    echo = new Echo({
      'broadcaster': 'pusher',
      'client': pusherClient,
      'auth': {
        'headers': {
          'Authorization':
              'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImp0aSI6IjFkNTcxYzcyMTcwZmFmNDRmM2JkYjdkZTE2NmVmNTMzZTMwYjYwNjI2MjA5NjJlNWFhMmNkNTkyOTgwMzNlZmVlMjMyNzI2YTczMTY2NGZiIn0.eyJhdWQiOiIxIiwianRpIjoiMWQ1NzFjNzIxNzBmYWY0NGYzYmRiN2RlMTY2ZWY1MzNlMzBiNjA2MjYyMDk2MmU1YWEyY2Q1OTI5ODAzM2VmZWUyMzI3MjZhNzMxNjY0ZmIiLCJpYXQiOjE1NTAyNTE4MzIsIm5iZiI6MTU1MDI1MTgzMiwiZXhwIjoxNTgxNzg3ODMyLCJzdWIiOiIxIiwic2NvcGVzIjpbXX0.Yhmzofhp8q32gcFmzQDAmm48cwmxqrZAi4Mfsbzp6pyzjzWziEM8isDrNqyQZANXUJP40LGJroK9LkVddHyqWtPQqTNXPv-Azx3tVy_YOkI-BhJhttKaiN7DTPD9gYiuUpINUjrqTVuzxzDHzXNmTemOVqVBABo4f6m9ZoWkdyNKyirPZHagofs4Lp1YR0--AgItZVCPl3DYrLdMY78a687edQaYA1q3HKWPggtxayN35BPoeOm08uxmE_YzYXRTzi6OxGLNcQnrC61NvWqTCZ5Ze4fvUuVRtASqybV1Y4oxcgs2kZYg1rxm5idHUrGwQE4fLmaGgDclBR7Ax-NbiN-VN1gJpB2D7eVYxigCVsCVBnoA5DtjKewg0axckthTXyxcTWJBnbarbbnPuI9ZWrxaZWcp4kWwosw9_pr7Ee_5Q-Uxr_ksZg6qZ3gVrqR9ZY0zZoQl4qEsbUGfLXLGrKj_NjeESOcUsPY6fACIy_Okttxto_a3y9wQwPMDJY-jxdLdZ1TdusZUJG8d5KJxfOUprqsCzju-TVnLZxpA-f15mI9zIcjtA-2PFYnUuukIxBw-FXzBWssrMLyTXxHl9Sux3WH5mtXDW_wbwoJj6aaLHBw28jgokbdlKx3QRWdTvAsALJM6nmDN4BWu2pIn05i5wpcUnxHIvcdOgoh3hn0'
        }
      }
    });
  }

  log(String event) {
    var now = new DateTime.now();
    String formatted =
        "${now.hour.toString().padLeft(2, '0')}:${now.minute.toString().padLeft(2, '0')}:${now.second.toString().padLeft(2, '0')}";

    setState(() {
      _logs.insert(0, "$formatted $event");
    });
  }

  void _listenToChannel(String type, String name, String event) {
    dynamic channel;

    if (type == 'public') {
      channel = echo.channel(name);
    } else if (type == 'private') {
      channel = echo.private(name);
    } else if (type == 'presence') {
      channel = echo.join(name).here((users) {
        print(users);
      }).joining((user) {
        print(user);
      }).leaving((user) {
        print(user);
      });
    }

    channel.listen(event, (e) {
      log('channel: $name, event: $event');
    });
  }

  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      child: Column(
        children: <Widget>[
          Flexible(
            child: Container(
              padding: EdgeInsets.all(15),
              color: Colors.grey[100],
              child: ListView.builder(
                reverse: true,
                itemCount: _logs.length,
                itemBuilder: (BuildContext context, index) {
                  return Padding(
                    padding: EdgeInsets.symmetric(vertical: 4),
                    child: Text(_logs[index]),
                  );
                },
              ),
            ),
          ),
          Container(
            height: 70,
            decoration: BoxDecoration(
              border: Border(
                top: BorderSide(color: Colors.grey[300]),
              ),
            ),
            child: Center(
              child: ListView(
                scrollDirection: Axis.horizontal,
                children: <Widget>[
                  CupertinoButton(
                    onPressed: () {
                      showCupertinoModalPopup<void>(
                        context: context,
                        builder: (BuildContext context) {
                          return ChannelModal(
                            listen: true,
                            type: channelType,
                            name: channelName,
                            onTypeChanged: (value) {
                              setState(() {
                                channelType = value;
                              });
                            },
                            onNameChanged: (value) {
                              setState(() {
                                channelName = value;
                              });
                            },
                            onEventChanged: (value) {
                              setState(() {
                                event = value;
                              });
                            },
                            onSubmit: () {
                              log('Listening to channel: $channelName');
                              _listenToChannel(channelType, channelName, event);
                              Navigator.of(context).pop();
                            },
                          );
                        },
                      );
                    },
                    child: Text('listen to channel'),
                  ),
                  CupertinoButton(
                    onPressed: () {
                      showCupertinoModalPopup<void>(
                        context: context,
                        builder: (BuildContext context) {
                          return ChannelModal(
                            listen: false,
                            name: channelName,
                            onNameChanged: (value) {
                              setState(() {
                                channelName = value;
                              });
                            },
                            onSubmit: () {
                              log('Leaving channel: $channelName');
                              echo.leave(channelName);
                              Navigator.of(context).pop();
                            },
                          );
                        },
                      );
                    },
                    child: Text('leave channel'),
                  ),
                  Visibility(
                    visible: !isConnected,
                    child: CupertinoButton(
                      onPressed: () {
                        log('connecting');
                        this.connect();
                      },
                      child: Text('connect'),
                    ),
                  ),
                  Visibility(
                    visible: isConnected,
                    child: CupertinoButton(
                      onPressed: () {
                        log('disconnecting');
                        echo.disconnect();
                      },
                      child: Text('disconnect'),
                    ),
                  ),
                  CupertinoButton(
                    onPressed: () {
                      dynamic id = echo.sockedId();
                      log('socket_id: $id');
                    },
                    child: Text('get socket-id'),
                  ),
                  CupertinoButton(
                    onPressed: () {
                      setState(() {
                        _logs = [];
                      });
                    },
                    child: Text('clear log'),
                  ),
                ],
              ),
            ),
          )
        ],
      ),
    );
  }
}
